expression/constant_propagation.go

```diff
@@ -138,10 +85,51 @@ func (s *propagateConstantSolver) propagateEQ() {
 	}
 }
 
+// propagateOthers propagates all deterministic conditions.
+// e.g. For expression a = b and b = c and c = d and c < 1 , we can get extra a < 1 and b < 1 and d < 1.
+// However, for a = b and a < rand(), we cannot propagate a < rand() to b < rand() because rand() is non-deterministic
+// We maintain a unionSet representing the equivalent for every two columns.
+func (s *propagateConstantSolver) propagateOthers() {
+	s.unionSet = &multiEqualSet{}
+	s.unionSet.init(len(s.columns))
+	for i := range s.conditions {
+		if fun, ok := s.conditions[i].(*ScalarFunction); ok && fun.FuncName.L == ast.EQ {
+			lCol, lOk := fun.GetArgs()[0].(*Column)
+			rCol, rOk := fun.GetArgs()[1].(*Column)
+			if lOk && rOk {
+				lID := s.getColID(lCol)
+				rID := s.getColID(rCol)
+				s.unionSet.addRelation(lID, rID)
+			}
+		}
+	}
+
+	condsLen := len(s.conditions)
+	for j, colj := range s.columns {
+		for k := j + 1; k < len(s.columns); k++ {
+			if s.unionSet.findRoot(j) != s.unionSet.findRoot(k) {
+				continue
+			}
+			colk := s.columns[k]
+			for i := 0; i < condsLen; i++ {
+				cond := s.conditions[i]
+				replaced, _, newExpr := s.tryToReplaceCond(colj, colk, cond)
+				if replaced {
+					s.conditions = append(s.conditions, newExpr)
+				}
+				replaced, _, newExpr = s.tryToReplaceCond(colk, colj, cond)
+				if replaced {
+					s.conditions = append(s.conditions, newExpr)
+				}
+			}
+		}
+	}
+}
+
 // validPropagateCond checks if the cond is an expression like [column op constant] and op is in the funNameMap.
-func (s *propagateConstantSolver) validPropagateCond(cond Expression, funNameMap map[string]bool) (*Column, *Constant) {
+func (s *propagateConstantSolver) validPropagateCond(cond Expression) (*Column, *Constant) {
 	if eq, ok := cond.(*ScalarFunction); ok {
-		if _, ok := funNameMap[eq.FuncName.L]; !ok {
+		if eq.FuncName.L != ast.EQ {
```

Comment:
1. update the comment of this function since we do not consider non-equal ops in this function.
2. s/ validPropagateCond/ validEqualCond?

Reply:
All right